datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -1225,12 +1228,33 @@ private void handleLeaderDoAssignment(boolean cleanUpOrphanNodes) {
         _adapter.updateAllAssignments(newAssignmentsByInstance);
       }
 
+      Set<String> failedStreams = Collections.emptySet();
+      // Poll the zookeeper to ensure that hosts claimed assignment tokens for stopping streams
+      if (_config.getEnableAssignmentTokens() &&
+          !PollUtils.poll(() -> _adapter.getNumUnclaimedTokensForDatastreams(stoppingDatastreamGroups) == 0,
```

Comment:
Let's listen to any exceptions that `updateAllAssignmentsAndIssueTokens` might throw and only perform this poll in the non-exception path, otherwise in worst-case scenarios, if the ZK is not updated with the tokens in `updateAllAssignmentsAndIssueTokens`, the count would always be 0, and this poll condition would meet.

Reply:
- If there's no tokens written to Zk (in the worst case scenario that you described), the poll will exit immediately. The condition for the if statement, however, won't be met, and the body of the if statement won't execute. This is expected behavior (no tokens were written, nothing to do here).
- In case of Zk create failure, I don't like the idea of handling Zk Exceptions in `handleLeaderDoAssignment`. We already have generic exception handling in `handleEvent` and it logs the exception messages. There's not much we can do in `handleLeaderDoAssignment`

Reply:
Link to exception handling code for reference:

https://github.com/linkedin/brooklin/blob/8bc670bd8bad20a6c2ff1fb335f3f3018c5b9b7b/datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java#L933